### PR TITLE
Isssue #34 

### PR DESCRIPTION
**Pull request by fredykoci**

### CODE REVIEW

File: src/client/commands/submit.rs [hunk 1]
```diff
@@ -12,6 +12,10 @@ use crate::transfer::connection::ClientConnection;
 use crate::transfer::messages::{FromClientMessage, JobType, SubmitRequest, ToClientMessage};
 use anyhow::anyhow;
 
+pub struct SubmitCommon {
```
Comment (spirali) [inline src/client/commands/submit.rs:15]: This is not needed

File: src/client/commands/submit.rs [hunk 2]
```diff
@@ -59,6 +68,7 @@ pub async fn submit_computation(
             stderr: Some(stderr),
             cwd: None,
         },
+        fail_no: opts.max_fails.unwrap(),
```
Comment (spirali) [inline src/client/commands/submit.rs:71]: * Please name the attributed and variables that hold this value as "max_fails"
* Do not call "unwrap" on data that came from the user. User should never cause "panic".
* If max fails is not defined, it should behave as the current version

File: src/server/job.rs [hunk 3]
```diff
@@ -45,13 +46,14 @@ impl JobTaskCounters {
 pub struct Job {
     pub job_id: JobId,
     pub base_task_id: TakoTaskId,
-
+    pub n_to_cancel_job: u32,
```
Comment (spirali) [inline src/server/job.rs:49]: Name it please max_fails

File: src/server/state.rs [hunk 4]
```diff
@@ -109,8 +109,19 @@ impl State {
...
+        if job.counters.n_failed_tasks >= job.n_to_cancel_job {
+            let test = job.make_job_detail(true);
+            for x in test.tasks.iter() {
+                let (_, state) = job.get_task_state_mut(x.task_id.into());
+                if matches!(state ,JobTaskState::Running) || matches!(state, JobTaskState::Waiting) {
+                    job.set_cancel_state(x.task_id.into());
+                }
+            }
+        } else {
+            job.set_failed_state(msg.id, msg.info.message);
```
Comment (spirali) [inline src/server/state.rs:123]: * ``job.set_failed_state(msg.id, msg.info.message);`` should be always called even max_fail was reached

**Review by spirali — CHANGES_REQUESTED** [review]

File: src/client/commands/submit.rs [hunk 5]
```diff
@@ -23,6 +24,10 @@ pub struct SubmitOpts {
 
     #[clap(long)]
     name: Option<String>,
+
+    #[clap(long, requires = "array")]
```
Comment (spirali) [inline src/client/commands/submit.rs:28]: Why is there requires = "array"?

File: src/server/state.rs [hunk 6]
```diff
@@ -40,6 +42,39 @@ pub type StateRef = WrappedRcRefCell<State>;
...
+    job_id: JobId,
+    tasks: Vec<TakoTaskId>,
+) {
+    if tasks.is_empty() {
+        return;
+    }
+    let tako_ref = tako_ref.clone();
+    let state_ref = state_ref.clone();
+    tokio::task::spawn_local(async move {
+        let message = FromGatewayMessage::CancelTasks(CancelTasks { tasks });
```
Comment (Kobzol) [inline src/server/state.rs:57]: `rpc_call!`?

Comment (spirali) [reply src/server/state.rs:57]: ``rpc_call`` is for client messages, this is a hq <-> tako message.

Comment (Kobzol) [reply src/server/state.rs:57]: Ah. So, should we introduce `tako_call!`? :)

Comment (spirali) [reply src/server/state.rs:57]: Good idea, but for another PR:)

**Pull request merged**